tapgarden/planter_test.go

```diff
@@ -1205,8 +1210,136 @@ func testFinalizeBatch(t *mintingTestHarness) {
 	t.assertNoPendingBatch()
 	t.assertNumCaretakersActive(caretakerCount)
 	t.assertLastBatchState(batchCount, tapgarden.BatchStateFinalized)
+}
+
+func testFinalizeWithTapscriptTree(t *mintingTestHarness) {
+	// First, create a new chain planter instance using the supplied test
+	// harness.
+	t.refreshChainPlanter()
+
+	// Create an initial batch of 5 seedlings.
+	const numSeedlings = 5
+	seedlings := t.queueInitialBatch(numSeedlings)
 
-	// TODO(jhb): add finalize with tapscript root
+	var (
+		wg          sync.WaitGroup
+		respChan    = make(chan *FinalizeBatchResp, 1)
+		finalizeReq tapgarden.FinalizeParams
+		batchCount  = 0
+	)
+
+	// Build a standalone tapscript tree object, that matches the tree
+	// created by other test helpers.
+	sigLockKey := test.RandPubKey(t)
+	hashLockWitness := []byte("foobar")
+	hashLockLeaf := test.ScriptHashLock(t.T, hashLockWitness)
+	sigLeaf := test.ScriptSchnorrSig(t.T, sigLockKey)
+	tapTreePreimage, err := asset.TapTreeNodesFromLeaves(
+		[]txscript.TapLeaf{hashLockLeaf, sigLeaf},
+	)
+	require.NoError(t, err)
+
+	finalizeReq = tapgarden.FinalizeParams{
+		TapTree: fn.Some(*tapTreePreimage),
+	}
+
+	// Force tapscript tree storage to fail, which should cause batch
+	// finalization to fail.
+	t.treeStore.FailStore = true
+	t.finalizeBatch(&wg, respChan, &finalizeReq)
+	finalizeErr := <-t.errChan
+	require.ErrorContains(t, finalizeErr, "unable to store tapscript tree")
+
+	// Allow tapscript tree storage to succeed, but force tapscript tree
+	// loading to fail.
+	t.treeStore.FailStore = false
+	t.treeStore.FailLoad = true
+	wg = sync.WaitGroup{}
```

Comment:
nit: wait group doesn't need explicit initialization.

Reply:
Agree, this was a workaround for the planter hanging after failing to freeze a batch - found a fix.

Reply:
Maybe was the issue of "attempted to copy mutex"?

Reply:
One branch of the planter logic on Finalize did not return an error to the caller, but only on `ErrChan` - so the caller would hang waiting for an error, but from the unit test context we can just read `ErrChan` and reset the waitGroup. Fixed the root cause in the planter, so this is no longer needed.